### ISSUE TITLE
Document returning filters from functions

### ISSUE THREAD

**Issue by blm768** [seanmonstar/warp#53]
I'm not really sure how much of this is an artifact of `warp`'s design and how much is just my lack of experience, but in a small test project, I found that attempting to abstract parts of filter/server construction in separate functions seems to produce some complex issues. For instance, consider the following reduced example:

```rust
extern crate warp;

pub fn get_routes() -> impl warp::Filter {
    warp::path::index()
}

fn main() {
    warp::serve(routes);
}
```

This produces the following error:

```
error[E0277]: the trait bound `<impl warp::Filter as warp::filter::FilterBase>::Extract: warp::reply::sealed::ReplySealed` is not satisfied
 --> src/main.rs:8:5
  |
8 |     warp::serve(get_routes());
  |     ^^^^^^^^^^^ the trait `warp::reply::sealed::ReplySealed` is not implemented for `<impl warp::Filter as warp::filter::FilterBase>::Extract`
  |
  = note: required because of the requirements on the impl of `warp::Reply` for `<impl warp::Filter as warp::filter::FilterBase>::Extract`
  = note: required because of the requirements on the impl of `warp::server::IntoWarpService` for `impl warp::Filter`
  = note: required by `warp::serve`
```

I haven't found a good workaround for the error yet; just extending the signature of `get_routes` to return `impl warp::Filter + Sync + Send + 'static` isn't enough to satisfy the bounds on the `impl` of `IntoWarpService` for `Filter`s in `src/filter/service.rs`.

Basically, it seems like there are so many trait bounds that trying to modularize a large application's route structure across multiple functions could get interesting fast. Most of the filter combinators (`or`, etc.) seem to include trait bounds that would cause similar issues to this one.

I don't really have a proposed solution (if this is even really a practical issue), but I guess it's something to think about…

**Comment by seanmonstar** [comment]
There's two ways to return a `Filter` from a function:

### With `impl Filter<Extract = T, Error = Rejection>`

```rust
pub fn index() -> impl warp::Filter<Extract = (), Error = Rejection> {
    warp::path::index()
}

pub fn index_ok() -> impl warp::Filter<Extract = (StatusCode,), Error = Rejection> {
    index()
        .map(|| StatusCode::OK)
}
```

### With `Filter::boxed()`

```rust
pub fn get_routes() -> BoxedFilter<(impl Reply,)> {
    some_filter()
        .or(some_other())
        .or(a_third)
        .with(my_log)
        // etc
        .boxed()
}
```

There's reasons you might pick one over the other.

- `impl Filter`
  - allows monomorphization, and so possibly better performance or optimizations.
- `BoxedFilter`
  - Uses dynamic dispatch, which can dramatic improve compile times.
  - You can't use nested `impl Trait`, so `impl Filter<Extract = impl Reply>` isn't legal Rust (yet). You can use with a concrete type, like `BoxedFilter<impl Reply>`.

**Comment by blm768** [comment]
That makes more sense. I'd looked at `BoxedFilter`, but I hadn't managed to figure out what type arguments to give it.
Do you think that this is a significant enough topic to have an example for it, or would it belong somewhere else in the documentation? I'd be happy to submit a PR with an example or doc updates.

**Comment by seanmonstar** [comment]
Yes, examples or docs are sorely needed. I've been meaning to flesh out the `warp::test` section of docs showing this off, but maybe there's better places. 

And unfortunately, since trait specialization isn't stable, it isn't immediately obvious that a filter extracting just 1 value needs to set `Extract = (Foo,)`, not `Extract = Foo`.

**Comment by blm768** [comment]
I can think of a couple of places in the docs that could be tweaked to give people with intermediate Rust experience enough to figure things out; an inline example of the correct type signature in `BoxedFilter` or `Filter::boxed()` might do it. I should be able to draft something up tomorrow.

**Comment by blm768** [comment]
Oh, wow. Just realized that `Filter::boxed()` already had examples. Apparently I need to learn how to read. But I've thrown together an example and put a basic example in the `BoxedFilter` docs, so I'll submit a PR.

**Comment by seanmonstar** [comment]
This was done, thanks!

**Comment by danbruder** [comment]
FYI for those of you coming to this thread - `impl Filter<Extract = impl Reply>` is allowed now!
